### PR TITLE
Native Fungible Token App

### PR DESCRIPTION
**Pull request by andresilva91**
## Motivation

We need a Native Fungible app

## Proposal

Implement a few system API calls:
- Created a `balance` query for the `fungible` app, so that both `fungible` and `native-fungible` share the same queries and mutations
- Renamed the `balance` function in `fungible` to `balance_or_default`, because it better describes what it does, and we needed a `balance` that would return None if the account wasn't found
- Created a native fungible app that reuses the fungible abi, has the same queries and mutations
    - Messages are disabled because all the messaging we need will happen on the system side
    - Sessions are not implemented, but will be implemented in a following PR
    - Application owners not supported yet, see follow up items
    - Ticker symbol on native fungible is fixed to `LIN`
- Created 3 contract system API calls: owner balance, transfer and claim
    - Transfer and claim internally call `SystemOperation`s
- And 1 service system API call: owner balance
- Reused and generalized existing `fungible` e2e tests to also test `native fungible`

## Test Plan

e2e tests
Tested fungible UI also

## Follow up items
- Implement support for sessions within native fungible
- Write simple README that will mostly point to the `fungible` README
- Double check that the `fungible` UI also works with `native fungible`, adapt what is needed
- Change codebase to use AccountOwner instead of Owner, and then as a follow up add support to Application owners to native fungible

### CODE REVIEW

**Comment by andresilva91** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #1675** <a href="https://app.graphite.dev/github/pr/linera-io/linera-protocol/1675?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/linera-io/linera-protocol/1675?utm_source=stack-comment).

File: examples/native-fungible/src/state.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+#[derive(RootView, async_graphql::SimpleObject)]
+// In the service we also want a metadata query, which is not derived from a struct member.
+// This attribute requires having a ComplexObject implementation that adds such fields.
+// The implementation with metadata is in service.rs. Since a ComplexObject impl is required,
+// there is also an empty one in contract.rs.
+#[graphql(complex)]
+#[view(context = "ViewStorageContext")]
+pub struct NativeFungibleToken {
+    // TODO(#968): We should support stateless Applications/empty user views
+    pub _dummy: RegisterView<u8>,
```
Comment (ma2bd) [inline examples/native-fungible/src/state.rs:17]: `RegisterView<()>` ?

File: linera-execution/src/execution_state_actor.rs [hunk 2]
```diff
@@ -97,6 +105,59 @@ where
                 callback.respond(balance);
             }
 
+            OwnerBalance { owner, callback } => {
+                let balance = *self.system.balances.get_mut_or_default(&owner).await?;
```
Comment (ma2bd) [inline linera-execution/src/execution_state_actor.rs:109]: `get(..)` should work

File: linera-execution/src/execution_state_actor.rs [hunk 3]
```diff
@@ -97,6 +105,59 @@ where
...
+            Transfer {
+                source,
+                destination,
+                amount,
+                callback,
+            } => {
+                let operation_context =
+                    operation_context.expect("Operation Context shouldn't be null here");
+
+                self.system
```
Comment (ma2bd) [inline linera-execution/src/execution_state_actor.rs:122]: Let's pass the authentication from the calling application inside the request and then create a fake `OperationContext` for now.

File: examples/native-fungible/src/state.rs (hunk 1 above)
Comment (andresilva91) [reply examples/native-fungible/src/state.rs:17]: I think to do that we need to implement graphql's `OutputType` for `()`, which from my understanding I don't think we can do :/ 
That's probably why we also use `u8` in the AMM

File: examples/native-fungible/src/contract.rs [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    /// Verifies that a transfer is authenticated for this local account.
+    fn check_account_authentication(
+        authenticated_signer: Option<Owner>,
+        owner: Option<Owner>,
+    ) -> Result<(), Error> {
+        if owner.is_some() && authenticated_signer == owner {
+            Ok(())
+        } else {
+            Err(Error::IncorrectAuthentication)
+        }
```
[the remark this answers is not in the document]
Comment (andresilva91) [reply examples/native-fungible/src/contract.rs:196]: Ah! I thought `ensure!` would panic ha. Will change it, way better, thank you!

File: examples/native-fungible/src/service.rs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        };
+        Ok(system_api::current_owner_balance(owner))
+    }
+}
+
+// Implements additional fields not derived from struct members of NativeFungibleToken.
+#[ComplexObject]
+impl NativeFungibleToken {
+    async fn ticker_symbol(&self) -> Result<String, async_graphql::Error> {
+        Ok(NativeFungibleToken::parameters()?.ticker_symbol)
```
[the remark this answers is not in the document]
Comment (andresilva91) [reply examples/native-fungible/src/service.rs:61]: Yes, that's something I'm intending to do still before marking this ready for review

File: linera-service/tests/end_to_end_tests.rs [hunk 6]
```diff
@@ -85,6 +85,38 @@ impl FungibleApp {
     }
 }
 
+struct NativeFungibleApp(FungibleApp);
```
[the remark this answers is not in the document]
Comment (andresilva91) [reply linera-service/tests/end_to_end_tests.rs:88]: So we can't because the balance query for the native fungible will be different. For the `FungibleApp` it just queries the existing state, we don't have state on the native fungible app

File: linera-service/tests/end_to_end_tests.rs [hunk 7]
```diff
@@ -562,6 +594,227 @@ async fn test_wasm_end_to_end_same_wallet_fungible(config: impl LineraNetConfig)
...
+        (account_owner1, Amount::ZERO),
+        (account_owner2, Amount::from_tokens(3)),
+    ])
+    .await;
+
+    node_service1.ensure_is_running().unwrap();
+    node_service2.ensure_is_running().unwrap();
+
+    net.ensure_is_running().await.unwrap();
+    net.terminate().await.unwrap();
```
[the remark this answers is not in the document]
Comment (andresilva91) [reply linera-service/tests/end_to_end_tests.rs:721]: It's not identical because we build the `native-fungible` example instead, use the `NativeFungibleApp` instead, the ticker will be fixed and we also have to `request_application` before calling `make_application` for chain2, because on the system API calls we won't register the application on chain2 when we transfer to it, like it happens for `fungible`.
We could try to parameterize all this, but I wouldn't be surprised if these tests eventually diverge, so didn't seem worth the effort. It would also be a lot of parameterization

File: linera-service/tests/end_to_end_tests.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (afck) [reply linera-service/tests/end_to_end_tests.rs:88]: But is that difference observable? Don't you make the same query, and get the same result?

File: linera-service/tests/end_to_end_tests.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (afck) [reply linera-service/tests/end_to_end_tests.rs:721]: On the other hand I think we want to keep them as compatible as possible. Not sure in how many places they would be different?

File: linera-service/tests/end_to_end_tests.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (andresilva91) [reply linera-service/tests/end_to_end_tests.rs:88]: The query for fungible is `accounts {{ entry(key: {}) {{ value }} }}`, because it queries the state map directly. For native fungible the query is `balance(owner: {})`, so it's different queries :/

File: linera-service/tests/end_to_end_tests.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (andresilva91) [reply linera-service/tests/end_to_end_tests.rs:721]: I see. Let me see if I can unify it

File: linera-service/tests/end_to_end_tests.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (afck) [reply linera-service/tests/end_to_end_tests.rs:88]: Ah, I see!
But we _could_ make the queries look identical, I think? And maybe we should?
We could allow both kinds of queries on both applications.

File: examples/native-fungible/src/service.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (andresilva91) [reply examples/native-fungible/src/service.rs:61]: One thing on this is that `Parameters` on the `FungibleAbi` expects a `ticker_symbol` as a parameter, so even if I require that it stays empty on initialization, I would still have to provide some string in the parameters when I create the application, right? Not sure how to get around that without altering the Abi

File: linera-service/tests/end_to_end_tests.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (andresilva91) [reply linera-service/tests/end_to_end_tests.rs:88]: True, I could do that. Let me look into it

File: examples/native-fungible/src/service.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (afck) [reply examples/native-fungible/src/service.rs:61]: Yes, it would just be required to be `""`. It's a compromise, since we want to have the same ABI.

[the remark this answers is not in the document]
Comment (andresilva91) [reply examples/native-fungible/src/service.rs:61]: Got it! Let me do that then

**Comment by andresilva91** [comment]
Updating just to make sure CI passes, but still have comments to address

File: linera-views-derive/src/snapshots/linera_views_derive__tests__test_generate_hash_view_code_C.snap [hunk 8]
```diff
@@ -36,3 +36,4 @@ where
         Ok(hasher.finalize())
     }
 }
+
```
Comment (andresilva91) [inline linera-views-derive/src/snapshots/linera_views_derive__tests__test_generate_hash_view_code_C.snap:39]: <i class='graphite__hidden'>[Re: line 40]<br/><br/></i>

I've ran `cargo insta test` locally, no errors, not sure why these are here 😅<span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/linera-io/linera-protocol/1675?utm_source=unchanged-line-comment">Graphite</a>.</span>

File: examples/crowd-funding/README.md [hunk 9]
```diff
@@ -268,13 +268,7 @@ check that we have received 110 tokens, in addition to the
...
 ```gql,uri=http://localhost:8080/chains/$CHAIN_0/applications/$APP_ID_0
 query {
-    accounts {
-        entry(
-            key: "User:7136460f0c87ae46f966f898d494c4b40c4ae8c527f4d1c0b1fa0f7cff91d20f"
-        ) {
-            value
-        }
-    }
+    balance(owner: "User:7136460f0c87ae46f966f898d494c4b40c4ae8c527f4d1c0b1fa0f7cff91d20f")
```
Comment (ma2bd) [inline examples/crowd-funding/README.md:271]: Afaict this README is still targeting the (non-native) fungible app, so this won't work.

**Review by ma2bd — COMMENTED** [review]
This PR is a bit large and reviewers could use some help:
* Can you move out the `insta` related changes?
* Can you summarize the changes in the "proposal" section?

File: linera-service/tests/end_to_end_tests.rs [hunk 10]
```diff
@@ -421,6 +430,10 @@ async fn test_wasm_end_to_end_fungible(config: impl LineraNetConfig) {
     ])
     .await;
 
+    node_service2
+        .request_application(&chain2, &application_id)
+        .await
+        .unwrap();
```
[the remark this answers is not in the document]
Comment (afck) [reply linera-service/tests/end_to_end_tests.rs:436]: It doesn't hurt to request the application, even when it's not necessary. I'd just keep it, to keep the code simpler, and maybe add a comment.

File: linera-execution/src/lib.rs [hunk 11]
```diff
@@ -334,6 +336,25 @@ pub trait BaseRuntime {
...
+        amount: Amount,
+    ) -> Result<(), ExecutionError>;
+
+    /// Claims amount from source to destination
+    fn claim(
+        &mut self,
+        source: Account,
+        destination: Account,
+        amount: Amount,
+    ) -> Result<(), ExecutionError>;
```
Comment (afck) [inline linera-execution/src/lib.rs:356]: Do `transfer` and `claim` really belong in `BaseRuntime`? They can't be called by a service, so I imagine they should be in `ContractRuntime`.

File: linera-execution/src/execution_state_actor.rs [hunk 12]
```diff
@@ -97,6 +101,63 @@ where
...
+                    .execute_operation(
+                        // Temporary hack creating fake OperationContext. We'll need just the
+                        // authenticated signer in the operation
+                        Self::get_fake_operation_context(signer),
+                        SystemOperation::Transfer {
+                            owner: source,
+                            recipient: crate::system::Recipient::Account(destination),
+                            amount,
+                            user_data: UserData::default(),
+                        },
```
Comment (afck) [inline linera-execution/src/execution_state_actor.rs:127]: Rather than creating a fake context and a `SystemOperation` that isn't actually in the block, maybe let's extract the `Transfer` part of `SystemExecutionStateView::execute_operation` into a separate method (`SystemExecutionStateView::transfer`?) instead.

Comment (andresilva91) [reply linera-execution/src/execution_state_actor.rs:127]: Ah, nice, sounds cleaner! Let me try

**Comment by ma2bd** [comment]
> Messages are disabled because all the messaging we need will happen on the system side.

After discussing with @afck , we could make the case for sending an empty message along transfers, so that the app gets auto-deployed. This is a hack but it gets the job done and we can revisit later.

File: examples/crowd-funding/README.md (hunk 9 above)
Comment (ma2bd) [reply examples/crowd-funding/README.md:271]: Oh nice!

Should we remove the (auto-generated?) `accounts` query in the fungible app?

Comment (andresilva91) [reply examples/crowd-funding/README.md:271]: Good idea! Let me try

**Comment by andresilva91** [comment]
Tested fungible UI as well, to make sure the `accounts` -> `balance` query changes didn't break it

File: linera-sdk/src/contract/conversions_to_wit.rs [hunk 13]
```diff
@@ -25,6 +25,28 @@ impl From<CryptoHash> for wit_system_api::CryptoHash {
...
+impl From<Owner> for wit_system_api::CryptoHash {
+    fn from(owner: Owner) -> Self {
+        let parts = <[u64; 4]>::from(owner.0);
+
+        wit_system_api::CryptoHash {
+            part1: parts[0],
+            part2: parts[1],
+            part3: parts[2],
+            part4: parts[3],
+        }
```
Comment (afck) [inline linera-sdk/src/contract/conversions_to_wit.rs:37]: Isn't that already implemented?

```suggestion
        wit_system_api::CryptoHash::from(owner.0)
```

File: linera-service/tests/end_to_end_tests.rs [hunk 14]
```diff
@@ -509,8 +523,13 @@ async fn test_wasm_end_to_end_same_wallet_fungible(config: impl LineraNetConfig)
     ]);
     let state = InitialState { accounts };
     // Setting up the application and verifying
-    let (contract, service) = client1.build_example("fungible").await.unwrap();
-    let params = fungible::Parameters::new("FUN");
+    let (contract, service) = client1.build_example(example_name).await.unwrap();
+    let params = if example_name == "native-fungible" {
+        // Native Fungible has a fixed LIN ticker symbol, so we don't need to set it
```
Comment (afck) [inline linera-service/tests/end_to_end_tests.rs:528]:
```suggestion
        // Native Fungible has a fixed NAT ticker symbol, so we don't need to set it
```

File: linera-service/tests/end_to_end_tests.rs [hunk 15]
```diff
@@ -381,8 +382,13 @@ async fn test_wasm_end_to_end_fungible(config: impl LineraNetConfig) {
...
```
Comment (afck) [inline linera-service/tests/end_to_end_tests.rs:387]:
```suggestion
        // Native Fungible has a fixed NAT ticker symbol, so we don't need to set it
```

File: examples/native-fungible/src/contract.rs [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+                owner,
+                amount,
+                destination,
+            } => {
+                Self::check_account_authentication(context.authenticated_signer, owner)?;
+                let owner = self.normalize_owner(owner);
+                let target_account = self.destination_to_account(destination);
+
+                system_api::transfer(Some(owner), target_account, amount);
+                Ok(ApplicationCallOutcome::default())
```
Comment (afck) [inline examples/native-fungible/src/contract.rs:162]: Shouldn't we also `get_transfer_outcome` here, and include the `Credit` message if it's for another chain?

File: examples/native-fungible/src/contract.rs [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+                Self::check_account_authentication(
+                    context.authenticated_signer,
+                    source_account.owner,
+                )?;
+
+                let source_account = self.normalize_account(source_account);
+                let target_account = self.normalize_account(target_account);
+
+                system_api::claim(source_account, target_account, amount);
+                Ok(ApplicationCallOutcome::default())
```
Comment (afck) [inline examples/native-fungible/src/contract.rs:179]: …and the `Claim` message.

File: examples/native-fungible/src/contract.rs [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+                target: _,
+                source: _,
+            } => Ok(ExecutionOutcome::default()),
+            Message::Withdraw {
+                owner,
+                amount,
+                target_account,
+            } => {
+                Self::check_account_authentication(context.authenticated_signer, owner)?;
+                Ok(self.get_transfer_outcome(owner, target_account, amount))
```
Comment (afck) [inline examples/native-fungible/src/contract.rs:128]: If the message is bounced, we should probably just ignore it.

File: examples/native-fungible/src/contract.rs [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+        }
+    }
+
+    fn normalize_account(&self, account: fungible::Account) -> Account {
+        let owner = self.normalize_owner(account.owner);
+        Account {
+            chain_id: account.chain_id,
+            owner: Some(owner),
+        }
+    }
```
Comment (afck) [inline examples/native-fungible/src/contract.rs:237]: Maybe we should (in `fungible` itself) `impl TryFrom<fungible::Account> for linera_sdk::base::Account`, and `impl From<linera_sdk::base::Account> for fungible::Account`?

File: linera-execution/src/execution_state_actor.rs [hunk 20]
```diff
@@ -97,6 +101,60 @@ where
...
+                signer,
+                callback,
+            } => {
+                let mut execution_outcome = RawExecutionOutcome::default();
+                let message = self
+                    .system
+                    .transfer(
+                        signer,
+                        source,
+                        crate::system::Recipient::Account(destination),
```
Comment (afck) [inline linera-execution/src/execution_state_actor.rs:122]: Why not import `Recipient`?

File: examples/native-fungible/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,4 @@
+// Copyright (c) Zefchain Labs, Inc.
```
Comment (afck) [inline examples/native-fungible/src/lib.rs:1]: Maybe we should add a short crate documentation, and use `cargo-rdme`, too.
I don't think it needs a full description like the other examples, just a link to `fungible` and an explanation of how it's different.

**Review by afck — APPROVED** [review]
Looks great! :+1: No blockers from my side, just a few nit-picks.

File: examples/crowd-funding/README.md (hunk 9 above)
Comment (ma2bd) [reply examples/crowd-funding/README.md:271]: Good job with the change. However, I'm having some afterthoughts on this: perhaps we got it backward and should instead implement the original GraphQL interface of `fungible` manually for `native-fungible`. Right now, the APIs are consistent but downgraded to a relatively naive version (which doesn't allow browsing the accounts for instance).

What do you think @Twey @afck?

Maybe we'd better leave the auto-generated API and add a TODO to support the same API in `native-fungible` and kill `balance`?

File: examples/fungible/src/lib.rs [hunk 22]
```diff
@@ -412,12 +412,9 @@ impl FungibleTokenAbi {
         chain: &ActiveChain,
         account_owner: AccountOwner,
     ) -> Option<Amount> {
-        let query = format!(
-            "query {{ accounts {{ entry(key: {}) {{ value }} }} }}",
-            account_owner.to_value()
-        );
+        let query = format!("query {{ balance(owner: {}) }}", account_owner.to_value());
```
Comment (ma2bd) [inline examples/fungible/src/lib.rs:415]: In the future, we could generate client code from GraphQL schemas and queries.

File: examples/native-fungible/src/contract.rs [hunk 23]
```diff
@@ -0,0 +1,287 @@
...
+    type Error = Error;
+    type Storage = ViewStateStorage<Self>;
+
+    async fn initialize(
+        &mut self,
+        _context: &OperationContext,
+        state: Self::InitializationArgument,
+    ) -> Result<ExecutionOutcome<Self::Message>, Self::Error> {
+        // Validate that the application parameters were configured correctly.
+        assert!(Self::parameters().is_ok());
```
[the remark this answers is not in the document]
Comment (ma2bd) [reply examples/native-fungible/src/contract.rs:37]: Failing is probably better than ignoring an inconsistent value, as long as the error message tells you what to do to fix the problem.

File: examples/native-fungible/src/contract.rs (hunk 18 above)
Comment (ma2bd) [reply examples/native-fungible/src/contract.rs:128]: Please let's get to the bottom of this. Did I forget this case in `fungible` as well?

File: linera-execution/src/lib.rs [hunk 24]
```diff
@@ -457,6 +462,22 @@ pub trait ContractRuntime: BaseRuntime {
     /// Consumes some of the execution fuel.
     fn consume_fuel(&mut self, fuel: u64) -> Result<(), ExecutionError>;
 
+    /// Transfers amount from source to destination
```
Comment (ma2bd) [inline linera-execution/src/lib.rs:465]: nit: Missing dot at the end (based on the rest of the file)

File: linera-service/tests/end_to_end_tests.rs [hunk 25]
```diff
@@ -381,8 +382,13 @@ async fn test_wasm_end_to_end_fungible(config: impl LineraNetConfig) {
...
+        fungible::Parameters::new("")
```
Comment (ma2bd) [inline linera-service/tests/end_to_end_tests.rs:388]: "NAT" (see other comment)

File: linera-service/tests/end_to_end_tests.rs [hunk 26]
```diff
@@ -41,13 +41,9 @@ struct FungibleApp(ApplicationWrapper<fungible::FungibleTokenAbi>);
 
 impl FungibleApp {
     async fn get_amount(&self, account_owner: &AccountOwner) -> Amount {
-        let query = format!(
-            "accounts {{ entry(key: {}) {{ value }} }}",
-            account_owner.to_value()
-        );
+        let query = format!("balance(owner: {})", account_owner.to_value());
```
Comment (ma2bd) [inline linera-service/tests/end_to_end_tests.rs:44]: If we plan to revert this changes, we should probably have in a separate PR (before this one)

File: examples/crowd-funding/README.md (hunk 9 above)
Comment (afck) [reply examples/crowd-funding/README.md:271]: Yes, I guess the service is less important to be compatible than the contract. I'm fine with leaving it as a TODO.

Alternatively, we could just support both kinds of query in `fungible`, and only the naive one in `native-fungible`?

File: examples/native-fungible/src/lib.rs (hunk 21 above)
Comment (andresilva91) [reply examples/native-fungible/src/lib.rs:1]: I'm doing that in the next PR with the README for this (it mostly points to `fungible`, like you said)

File: examples/crowd-funding/README.md (hunk 9 above)
Comment (andresilva91) [reply examples/crowd-funding/README.md:271]: As discussed during co-working time, will add `accounts` support for native fungible in a following PR, and keep using that everywhere

File: examples/native-fungible/src/contract.rs (hunk 18 above)
Comment (andresilva91) [reply examples/native-fungible/src/contract.rs:128]: As discussed in the meeting, since we're doing nothing here it doesn't matter if we're not taking bouncing credit messages into account. I'll add a comment though

File: linera-service/tests/end_to_end_tests.rs (hunk 26 above)
Comment (andresilva91) [reply linera-service/tests/end_to_end_tests.rs:44]: Gonna add `accounts` support to native fungible in a following PR

File: examples/native-fungible/src/contract.rs [hunk 27]
```diff
@@ -0,0 +1,323 @@
...
```
Comment (afck) [inline examples/native-fungible/src/contract.rs:37]: This line is redundant now; the next one seems to also check whether it's `Ok`.)

File: examples/native-fungible/src/contract.rs [hunk 28]
```diff
@@ -0,0 +1,323 @@
...
+                // to check if it's a bouncing message like in the fungible app
+                Ok(ExecutionOutcome::default())
+            }
+            Message::Withdraw {
+                owner,
+                amount,
+                target_account,
+            } => {
+                Self::check_account_authentication(context.authenticated_signer, owner)?;
+                Ok(self.get_transfer_outcome(owner, target_account, amount))
```
Comment (afck) [inline examples/native-fungible/src/contract.rs:126]: I know the message doesn't do anything, but that `Credit` message we're creating here is really misleading if this message is bounced, isn't it?
Doesn't have to be in this PR, but I really think we shouldn't send it if the `Withdraw` was bounced.

File: linera-execution/src/system.rs [hunk 29]
```diff
@@ -718,6 +685,81 @@ where
...
+            SystemExecutionError::IncorrectTransferAmount
+        );
+        let balance = match &owner {
+            Some(owner) => self.balances.get_mut_or_default(owner).await?,
+            None => self.balance.get_mut(),
+        };
+        balance
+            .try_sub_assign(amount)
+            .map_err(|_| SystemExecutionError::InsufficientFunding { balance: *balance })?;
+        if let Recipient::Account(account) = recipient {
```
Comment (afck) [inline linera-execution/src/system.rs:712]: It would be safer to use `match` here, without a `_` arm. So if we ever add another account type, we don't accidentally burn the tokens here.

**Review by afck — APPROVED** [review]

File: examples/native-fungible/src/contract.rs (hunk 28 above)
Comment (afck) [reply examples/native-fungible/src/contract.rs:126]: (As discussed, if they aren't tracked, they will never be bounced.)
